### PR TITLE
Splitting the flatcc runtime dependency into parsing and building.

### PR DESCRIPTION
**Pull request by benvanik**
Even though LTO will strip what we don't use it still means we are
compiling more than we need and risking inclusion of things we don't want.
I still don't like how the compiler depends on the runtime flatcc files
but that's a task for another day.

### CODE REVIEW

**Review by hcindyl — CHANGES_REQUESTED** [review]
Not sure if we have covered everything.The cross-compile presubmit still shows flatcc runtime builder. Could you please double check?

```

[107/155] Building C object build_tools/third_party/flatcc/CMakeFiles/flatcc_runtime.dir/__/__/__/third_party/flatcc/src/runtime/builder.c.obj
[108/155] Building C object iree/vm/CMakeFiles/iree_vm_impl.dir/shims.c.obj
[109/155] Building C object build_tools/third_party/flatcc/CMakeFiles/flatcc_runtime.dir/__/__/__/third_party/flatcc/src/runtime/json_parser.c.obj
[110/155] Linking C static library iree/vm/libiree_vm_impl.a
[111/155] Linking C static library iree/modules/vmvx/libiree_modules_vmvx_vmvx.a
[112/155] Building C object iree/hal/local/loaders/CMakeFiles/iree_hal_local_loaders_system_library_loader.dir/system_library_loader.c.obj
[113/155] Building C object iree/modules/hal/CMakeFiles/iree_modules_hal_hal.dir/module.c.obj
[114/155] Linking C static library iree/modules/hal/libiree_modules_hal_hal.a
[115/155] Building C object build_tools/third_party/flatcc/CMakeFiles/flatcc_runtime.dir/__/__/__/third_party/flatcc/src/runtime/json_printer.c.obj
[116/155] Building C object iree/vm/CMakeFiles/iree_vm_bytecode_module.dir/bytecode_module.c.obj
[117/155] Linking C static library build_tools/third_party/flatcc/libflatcc_runtime.a
```

**Comment by hcindyl** [comment]
https://gist.github.com/hcindyl/dd72759018fc3d414ff22343921ed969

rv32 baremetal target build.ninja snippet with `libflatcc_runtime.a` dependency based on this PR.

**Comment by benvanik** [comment]
Darn, will take a look. Good idea checking the ninja output!

File: build_tools/third_party/flatcc/BUILD.overlay [hunk 1]
```diff
@@ -15,7 +15,6 @@ cc_library(
         "src/runtime/json_parser.c",
         "src/runtime/json_printer.c",
         "src/runtime/refmap.c",
-        "src/runtime/verifier.c",
     ],
     hdrs = [
         "include/flatcc/flatcc_accessors.h",
```
[the remark this answers is not in the document]
Comment (benvanik) [reply build_tools/third_party/flatcc/BUILD.overlay:20]: unfortunately transitive hdrs don't work in bazel, so if someone deps on :runtime to get the whole runtime we don't also want them to dep on :parsing where the hdrs are declared - the important part is we don't duplicate the srcs so we don't trigger ODR violations, but the headers are fine (basically just declares "you can #include this header by depending on this library")

**Comment by benvanik** [comment]
Ok think I fixed it - no more flatcc_runtime dep from simple_embedding_test!

**Comment by hcindyl** [comment]
The RV64 Linux and ARM64 Android runtimes still build libflatcc_runtime.a and flatcc/runtime/builder.c and flatcc/runtime/json_printer.c objects.
https://source.cloud.google.com/results/invocations/a1ed384c-062d-4d36-a3e9-9033318e8766/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Friscv64%2Fpresubmit/log

https://source.cloud.google.com/results/invocations/bd2f0a0a-55c4-4dbb-a82b-4ed66ee30a0c/targets/iree%2Fgcp_ubuntu%2Fcmake%2Fandroid%2Farm64-v8a%2Fpresubmit/log

The dependency is not shown in rv32 baremetal, so we could address that in a later PR if you want unblock baremetal first.

**Comment by benvanik** [comment]
I think (but could be wrong a second time ;) that those are because that build is producing `Linking CXX executable iree/tools/iree-dump-module` - but the non-tool files shouldn't be requiring it.

**Review by hcindyl — APPROVED** [review]

**Pull request merged**